Z-levels: bashing, tiny bit of mapgen

Coolthulhu opened the pull request:
This time I didn't create separate overloads, but changed the existing functions. I used C++11's `= delete` to make it less likely that a function with default arguments will cast `int` to `const tripoint &p`.

Things I changed and tested here:
- Opening and closing doors
- Bashing stuff
- Shooting at and through stuff (found a bug in master - shooting through glass doors never breaks them)
- Exploding stuff
- Wrecking entire houses (found a bug in master - totally devastated houses can have hanging ceilings with no support)
- Car vs car collisions, car vs building collisions, car vs creature collisions (found two rare bugs in master - some really powerful car vs car collisions can result in "Part already removed" debugmsg and sometimes this spawns wreckage far from the collision spot)
- Car vs player collision (minor bug in master: doesn't cause boarding, even when the player is not knocked back)
- Sign generation, player altering signs
- Computers being generated, used to open doors and glass
- Rubble generation
- CBM generation from butchery

By "bugs in master" I mean "I managed to replicate it with the newest binary".
Didn't fix the bugs in master. The one related to what I'm doing here is the house wrecking one, but it happens only rarely (actually, all of the bugs here are rare) and will need to be reworked later on (Z-level buildings), so I didn't search for what causes it.

DavidKeaton commented:
Do you have an overview issue with your todos?

Coolthulhu commented:
At the moment no.
I'll be changing light/vision related stuff next, then probably removing 2D overloads starting from the top of map.h.

BevapDin commented:
Regarding mapgen: currently mapgen is only done on a single z-level, is it planed at make it work on multiple z-levels at once? If not, the mapgen functions (and `map` functions like `map::draw_rough_circle_furn`) won't need the z-parameter and you won't need to change all the code in mapgen.cpp.

Generating *all* z-levels at once might be a bit of performance problem, I already experiences short lag on this (especially when creating a vehicle wreckage).

Coolthulhu commented:
Unfortunately all z-levels will have to be generated for a single tile to be loaded. Going by the discussion in #6818, all z-levels of the bubble are supposed to be loaded at all times.

Loading/generation may be spread out somehow (for example, adding a second process to do it in the background before they get in the bubble), but it has to be done to all 21 z-levels more or less at once.

Generating all 21 z-levels is noticeably slower, but it is optimized for empty levels so most of the time it only actually generates 2-3 of them.

Now that I think about it, I'll probably leave some of the 2D overloads just for mapgen, because there are tons of instances of their usage there. Adding a `tripoint( x, y, abs_sub.z )` wrapper to all of them (or even just a third argument) would be a lot of useless code changes.
I could try making them private and friend them just for the mapgen. All other code could use the 3D ones.

DavidKeaton commented:
If the player is walking on ground level, there would be less need to do z processing for some N layers below them, maybe less than that above him? (Air vs. Dirt)

Coolthulhu commented:
Generating pure air and pure rock submaps is cheap - it is cheap enough that there was a major performance gain when I made it so that they're re-generated every time rather than saved.
The problem is, there still are ~18 of those to generate and 2-3 non-empty ones to either generate or load.

DavidKeaton commented:
For the pure submaps, if nothing differs, why not flag it as such and just put a reference to the already generated one? Objects can still be placed on it with coordinates, but on a "shadow" layer? As in, Z >= 4 are the same, all Z > 4 are reference to 4, but can still draw a bird on Z = 6 by doing a copy sym  of 4 then draw bird over that?

Coolthulhu commented:
Generating itself is really quick (just 5 memsets) and memory isn't a problem, the problem is that all the setup before generation.
For every 4 submaps (1 overmap square) to be generated, there's a separate empty map created just for those 4 submaps, all neighbors are checked for monster spawns, all neighbors are checked for types (even if they're to be discarded) etc.

I'll probably add a special mapgen function that skips all that, just so that empty submaps don't need to do all that.

DavidKeaton commented:
You could theoretically even leave pure submaps to static initialization, and just reference them there. Less generation = less overhead! :-)

I would even suggest relative changes to a submap of almost exclusively terrain/whatevs of type X, but I'm just a ramblin' man.

KA101 commented:
@vache, any issues with your computer reworks?  If not, let me know and I'll see about

Coolthulhu commented:
From `computer.cpp`, I changed coordinates in `translate_radius` (used to open doors), `make_rubble` and `explosion`. Computer-spawning function (used in mapgen) was also changed.

Other than that, computer stuff should be unaffected.

kevingranade commented:
BTW, I'm updating #2386, the goal of which is to make lighting and vision range consistent and straightforward instead of basing it on a bunch of heuristics.  Basically "open air" will not be totally transparent, the level of cumulative opacity will be calculated for each tile in FoV, and used as a scaling factor for lighting, if apparent lighting is lower than some threshold, that will make that square not visible.  The initial values will be chosen to match the current visibility level, i.e. in full light visibility will fall off to 0 after 60 squares, and lower light levels will drop visibility.  The difference is it will provide a single system for handling partial opacity, e.g. smoke and similar, possibly even tinted glass or the like.

Coolthulhu commented:
How long till the lightning update is out? I'd only change the function declarations and first few lines, but if it would be more than about a week, I'll just set up overloads instead when updating map.h.

vache commented:
This all looks pretty compatible with all the computer stuff I have now.  I'll have to update the computer actions I've written to include z-levels now though.  Don't hold anything back on my account though for sure.

kevingranade commented:
The damn shadowcasting code is really hard to work with, but I think I have the hard part sorted out now.
My current progress is at #11935
Isn't updating the shadowcasting code a separate bounty?  Are you going after them all at once?

Coolthulhu commented:
No, I want to do them roughly in order, but the Part 1 is rather vague on what should be included and what not, so I thought I'm supposed to add at least the overloads for FoV functions.

The way I see it, Part 1 is done when (all? most?) `map` class functions can be accessed with 3D coordinates and the most basic, uncached functions (ones explicitly mentioned in the Part 1 issue, like `ter_at`) produce correct results outside current z-level.

DavidKeaton commented on src/map.h:
```diff
@@ -578,50 +581,54 @@ void draw_rough_circle(std::string type, int x, int y, int rad);
+    std::pair<bool, bool> bash( const tripoint &p, const int str, bool silent = false,
+                                bool destroy = false, vehicle *bashing_vehicle = nullptr );
+    std::pair<bool, bool> bash( int, int, const int, bool silent = false,
+                                bool destroy = false, vehicle *bashing_vehicle = nullptr ) = delete;
```
C++ is still something I am learning the more I contribute, so I have a question.
I read the open standards draft regarding `= delete` when I saw this here, but want to make sure I got this right.
In this context, you are disallowing use of the old 2D coordinate versions of the functions? Or is it so it does not incorrectly copy the contents over to tripoint type? (I know it can be used to disallow overloaded conversions, or some jazz).

Coolthulhu replied:
Yeah. I made the error of thinking compiler would be smart enough not to convert `int` to `const tripoint&`, but apparently to the compiler it's not even worth a warning.

Has resulted in some quite weird bugs few versions ago and explicitly deleting the function is safer than manually double-checking. Plus, it protects other contributors from using the just-removed overload.

DavidKeaton replied:
so it can be used as sorta of a way of saying "I removed this function, don't use it" cause of the compiler error if trying to use said deleted function, yeah?

Coolthulhu replied:
Yeah. If you try to compile a file that uses a deleted function, the compiler will not attempt implicit casts and just throw an error.
If I could, I'd just delete the `int`->`tripoint&` operator, but I'm not sure if it can be done.

DavidKeaton replied:
cool! thanks!

BevapDin replied:
> Yeah. Few versions ago I made the error of thinking compiler would be smart enough not to convert  int  to  const tripoint& , but apparently to the compiler it's not even worth a warning. Has resulted in some quite weird bugs when it happened.

Well, the compiler is right. There is a (non-explicit) tripoint constructor that accepts exactly one `int` as parameter: `tripoint(int X = 0, int Y = 0, int Z = 0)` - Y and Z get the default value and you have a constructor that looks like `tripoint(int X)`, which looks like a suitable way to convert an `int` to a `tripoint`.

I suggest removing the default parameters there, where are they every used anyway?

Coolthulhu replied:
I guess I'm too used to the high-level languages - I didn't think about the possibility of it constructing a new tripoint from that int.

Yeah, the default arguments can be safely removed. They seem to be only used in one place in the code.

kevingranade replied:
I'm guessing it's used for declaring a null tripoint, in that case I'd say
just use an explicit no-argument constructor (which we might have?).

Coolthulhu replied:
We have the `tripoint( int, int, int )` with 3 default arguments. In #11972 I remove the default arguments and add an explicit no-argument one.
`point` class still has an overload that takes 1 argument, though.
